File: ScapesCore/src/main/java/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteChunkManagerClient.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.chunk.terrain.infinite;

import java8.util.Objects;
import java8.util.Optional;
import java8.util.stream.Collectors;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.FastMath;

import java.util.Arrays;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicLong;

public class TerrainInfiniteChunkManagerClient
        implements TerrainInfiniteChunkManager {
    private final int radius, size;
    private final TerrainInfiniteChunkClient[] array;
    private final AtomicLong lock = new AtomicLong();
    private int x, y;

    public TerrainInfiniteChunkManagerClient(int radius) {
        this.radius = radius;
        size = (radius << 1) + 1;
        array = new TerrainInfiniteChunkClient[size * size];
    }

    public synchronized void add(TerrainInfiniteChunkClient chunk) {
        int xx = chunk.x() - x;
        int yy = chunk.y() - y;
        if (xx >= 0 && xx < size && yy >= 0 && yy < size) {
            int i = yy * size + xx;
            array[i] = null;
            array[i] = chunk;
        }
    }

    public synchronized Optional<TerrainInfiniteChunkClient> remove(int x,
            int y) {
        int xx = x - this.x;
        int yy = y - this.y;
        if (xx >= 0 && xx < size && yy >= 0 && yy < size) {
            int i = yy * size + xx;
            TerrainInfiniteChunkClient chunk = array[i];
            if (chunk != null) {
                array[i] = null;
                return Optional.of(chunk);
            }
        }
        return Optional.empty();
    }

    @Override
    public Optional<TerrainInfiniteChunkClient> get(int x, int y) {
        long stamp = lock.get();
        int xx = x - this.x;
        int yy = y - this.y;
        if (xx >= 0 && xx < size && yy >= 0 && yy < size) {
            int i = yy * size + xx;
            TerrainInfiniteChunkClient value = array[i];
            long validate = lock.get();
            if (stamp == validate && (validate & 1) == 0) {
                if (value == null) {
                    return Optional.empty();
                } else {
                    return value.optional();
                }
            }
        }
        synchronized (this) {
            xx = x - this.x;
            yy = y - this.y;
            if (xx >= 0 && xx < size && yy >= 0 && yy < size) {
                int i = yy * size + xx;
                TerrainInfiniteChunkClient value = array[i];
                if (value == null) {
                    return Optional.empty();
                } else {
                    return value.optional();
                }
            } else {
                return Optional.empty();
            }
        }
    }

    @Override
    public boolean has(int x, int y) {
        return get(x, y) != null;
    }

    @Override
    public Collection<TerrainInfiniteChunkClient> iterator() {
        return Streams.of(array).filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    protected boolean setCenter(int x, int y) {
        x -= radius;
        y -= radius;
        if (x != this.x || y != this.y) {
            synchronized (this) {
                lock.incrementAndGet();
                int xDiff = this.x - x;
                int xDiffAbs = FastMath.abs(xDiff);
                if (xDiffAbs > 0) {
                    if (xDiffAbs > size) {
                        clear();
                    } else {
                        for (int i = 0; i < xDiff; i++) {
                            shiftXPositive();
                        }
                        xDiff = -xDiff;
                        for (int i = 0; i < xDiff; i++) {
                            shiftXNegative();
                        }
                    }
                    this.x = x;
                }
                int yDiff = this.y - y;
                int yDiffAbs = FastMath.abs(yDiff);
                if (yDiffAbs > 0) {
                    if (yDiffAbs > size) {
                        clear();
                    } else {
                        for (int i = 0; i < yDiff; i++) {
                            shiftYPositive();
                        }
                        yDiff = -yDiff;
                        for (int i = 0; i < yDiff; i++) {
                            shiftYNegative();
                        }
                    }
                    this.y = y;
                }
                lock.incrementAndGet();
            }
            return true;
        }
        return false;
    }

    private void clear() {
        Arrays.fill(array, null);
    }

    private void shiftXPositive() {
        System.arraycopy(array, 0, array, 1, array.length - 1);
        for (int i = 0; i < array.length; i += size) {
            array[i] = null;
        }
    }

    private void shiftXNegative() {
        System.arraycopy(array, 1, array, 0, array.length - 1);
        for (int i = size - 1; i < array.length; i += size) {
            array[i] = null;
        }
    }

    private void shiftYPositive() {
        System.arraycopy(array, 0, array, size, array.length - size);
        for (int i = 0; i < size; i++) {
            array[i] = null;
        }
    }

    private void shiftYNegative() {
        System.arraycopy(array, size, array, 0, array.length - size);
        for (int i = array.length - size; i < array.length; i++) {
            array[i] = null;
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteChunkManagerClient.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.chunk.terrain.infinite

import org.tobi29.scapes.engine.utils.StampLock
import org.tobi29.scapes.engine.utils.math.abs
import org.tobi29.scapes.entity.client.EntityClient
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class TerrainInfiniteChunkManagerClient(private val radius: Int) : TerrainInfiniteChunkManager<EntityClient> {
    private val size: Int
    private val array: Array<TerrainInfiniteChunkClient?>
    private val lock = StampLock()
    private val x = AtomicInteger()
    private val y = AtomicInteger()

    init {
        size = (radius shl 1) + 1
        array = arrayOfNulls<TerrainInfiniteChunkClient>(size * size)
    }

    fun add(chunk: TerrainInfiniteChunkClient) {
        lock.write {
            val xx = chunk.pos.x - x.get()
            val yy = chunk.pos.y - y.get()
            if (xx in 0..(size - 1) && yy >= 0 && yy < size) {
                val i = yy * size + xx
                array[i] = chunk
            }
        }
    }

    fun remove(x: Int,
               y: Int): TerrainInfiniteChunkClient? {
        lock.write {
            val xx = x - this.x.get()
            val yy = y - this.y.get()
            if (xx in 0..(size - 1) && yy >= 0 && yy < size) {
                val i = yy * size + xx
                val chunk = array[i]
                if (chunk != null) {
                    assert(chunk.pos.x == x)
                    assert(chunk.pos.y == y)
                    array[i] = null
                    return chunk
                }
            }
            return null
        }
    }

    override fun get(x: Int,
                     y: Int): TerrainInfiniteChunkClient? {
        val value = lock.read {
            val xx = x - this.x.get()
            val yy = y - this.y.get()
            if (xx in 0..(size - 1) && yy >= 0 && yy < size) {
                val i = yy * size + xx
                array[i]
            } else {
                null
            }
        }
        value?.let {
            assert(value.pos.x == x)
            assert(value.pos.y == y)
        }
        return value
    }

    override fun has(x: Int,
                     y: Int): Boolean {
        return get(x, y) != null
    }

    override fun stream(): Sequence<TerrainInfiniteChunkClient> {
        return array.asSequence().filterNotNull()
    }

    internal fun setCenter(x: Int,
                           y: Int,
                           removed: (TerrainInfiniteChunkClient) -> Unit): Boolean {
        val xx = x - radius
        val yy = y - radius
        if (xx != this.x.get() || yy != this.y.get()) {
            lock.write {
                var xDiff = this.x.get() - xx
                val xDiffAbs = abs(xDiff)
                if (xDiffAbs > 0) {
                    if (xDiffAbs > size) {
                        clear()
                    } else {
                        for (i in 0..xDiff - 1) {
                            shiftXPositive(removed)
                        }
                        xDiff = -xDiff
                        for (i in 0..xDiff - 1) {
                            shiftXNegative(removed)
                        }
                    }
                    this.x.set(xx)
                }
                var yDiff = this.y.get() - yy
                val yDiffAbs = abs(yDiff)
                if (yDiffAbs > 0) {
                    if (yDiffAbs > size) {
                        clear()
                    } else {
                        for (i in 0..yDiff - 1) {
                            shiftYPositive(removed)
                        }
                        yDiff = -yDiff
                        for (i in 0..yDiff - 1) {
                            shiftYNegative(removed)
                        }
                    }
                    this.y.set(yy)
                }
            }
            return true
        }
        return false
    }

    private fun clear() {
        Arrays.fill(array, null)
    }

    private fun shiftXPositive(removed: (TerrainInfiniteChunkClient) -> Unit) {
        var i = 0
        while (i < array.size) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
            i += size
        }
        i = size - 1
        while (i < array.size) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
            i += size
        }
        System.arraycopy(array, 0, array, 1, array.size - 1)
    }

    private fun shiftXNegative(removed: (TerrainInfiniteChunkClient) -> Unit) {
        var i = 0
        while (i < array.size) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
            i += size
        }
        i = size - 1
        while (i < array.size) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
            i += size
        }
        System.arraycopy(array, 1, array, 0, array.size - 1)
    }

    private fun shiftYPositive(removed: (TerrainInfiniteChunkClient) -> Unit) {
        for (i in 0..size - 1) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
        }
        for (i in array.size - size..array.size - 1) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
        }
        System.arraycopy(array, 0, array, size, array.size - size)
    }

    private fun shiftYNegative(removed: (TerrainInfiniteChunkClient) -> Unit) {
        for (i in 0..size - 1) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
        }
        for (i in array.size - size..array.size - 1) {
            val chunk = array[i]
            chunk?.let { removed(it); array[i] = null }
        }
        System.arraycopy(array, size, array, 0, array.size - size)
    }
}
